### PR TITLE
feat(OAuth): add OAuth audience parameter support for Atlassian and Auth0

### LINKED ISSUE #3499: [EPIC][AUTH]: OAuth `audience` parameter support — enable Atlassian and non-RFC-8707 OAuth providers
# Epic: OAuth `audience` Parameter Support

## Goal

Add an optional `audience` parameter to the gateway's OAuth configuration so that providers requiring `audience` instead of (or in addition to) the RFC 8707 `resource` parameter — such as **Atlassian**, **Auth0**, and others — can be integrated without workarounds.

## Why Now?

1. **Atlassian MCP integration is blocked**: Atlassian's OAuth endpoint requires `audience=api.atlassian.com` but ContextForge only sends `resource` (RFC 8707). Users cannot authorize Atlassian MCP servers through the gateway.
2. **Auth0 uses `audience` natively**: Auth0's Authorization API requires `audience` to scope tokens — a common enterprise IdP.
3. **Entra v2 precedent exists**: The codebase already has special handling to omit `resource` for Microsoft Entra v2 endpoints (`_should_include_resource_parameter()`), demonstrating that provider-specific parameter handling is an established pattern.
4. **Growing adoption**: Multiple community members have reported OAuth integration friction with non-RFC-8707 providers.

---

## 📖 User Stories

<details>
<summary><strong>US-1: Integration Engineer — Authorize Atlassian MCP Server</strong></summary>

**As an** Integration Engineer
**I want** to configure an `audience` parameter in the gateway's OAuth settings
**So that** I can connect Atlassian MCP servers through the gateway

**Acceptance Criteria:**

```gherkin
Given a gateway is configured with:
  oauth_config:
    grant_type: authorization_code
    client_id: "my-atlassian-app"
    authorization_url: "https://auth.atlassian.com/authorize"
    token_url: "https://auth.atlassian.com/oauth/token"
    audience: "api.atlassian.com"
    scopes: ["read:jira-work", "write:jira-work"]
When the admin clicks "Authorize" in the Admin UI
Then the authorization URL should include `audience=api.atlassian.com`
And the token exchange request should include `audience=api.atlassian.com`
And the `resource` parameter should NOT be sent (since `audience` is set)
And the OAuth flow should complete successfully
```

</details>

<details>
<summary><strong>US-2: Platform Admin — Configure Auth0 OAuth with Audience</strong></summary>

**As a** Platform Administrator
**I want** to use Auth0 as my OAuth provider with audience-scoped tokens
**So that** tokens issued to the gateway are properly scoped to my API

**Acceptance Criteria:**

```gherkin
Given a gateway is configured with:
  oauth_config:
    grant_type: client_credentials
    client_id: "my-auth0-client"
    client_secret: "encrypted-secret"
    token_url: "https://my-tenant.auth0.com/oauth/token"
    audience: "https://my-api.example.com"
When the gateway requests a token for tool invocation
Then the token request should include `audience=https://my-api.example.com`
And the returned token should be scoped to that audience
```

</details>

<details>
<summary><strong>US-3: Gateway Admin — Use Both Resource and Audience</strong></summary>

**As a** Gateway Administrator
**I want** to configure both `resource` and `audience` for providers that accept both
**So that** I can comply with RFC 8707 while also satisfying provider-specific requirements

**Acceptance Criteria:**

```gherkin
Given a gateway oauth_config has both `resource` and `audience` set
When an authorization or token exchange request is made
Then both parameters should be included in the request
And neither should override the other
```

</details>

---

### Affected Files

| File | Change |
|------|--------|
| `mcpgateway/services/oauth_manager.py` | Add `audience` to auth URL construction (line ~1150) and token exchange (line ~1200) |
| `mcpgateway/services/oauth_manager.py` | Update `_should_include_resource_parameter()` to handle `audience`-only configs |
| `mcpgateway/schemas.py` | Document `audience` as a recognized `oauth_config` key (multiple schema locations) |
| `mcpgateway/admin.py` | Add `audience` field to the OAuth configuration form in Admin UI |
| `docs/docs/manage/` | Update OAuth configuration documentation |

### OAuth Flow with `audience`

```mermaid
sequenceDiagram
    participant Admin as Admin UI
    participant GW as Gateway
    participant AS as Auth Server (Atlassian/Auth0)

    Admin->>GW: Configure oauth_config with audience
    Admin->>GW: Click "Authorize"
    GW->>GW: _create_authorization_url_with_pkce()
    Note over GW: Include audience param if set
    GW->>AS: GET /authorize?audience=api.atlassian.com&...
    AS-->>Admin: Consent screen
    Admin->>AS: Approve
    AS->>GW: Callback with auth code
    GW->>GW: _exchange_code_for_tokens()
    Note over GW: Include audience in token request
    GW->>AS: POST /token {audience, code, ...}
    AS-->>GW: Access token + refresh token
    GW->>GW: Store tokens (encrypted)
```

### Parameter Precedence Logic

```
if oauth_config.audience is set:
    include audience in auth URL and token exchange
if oauth_config.resource is set AND _should_include_resource_parameter():
    include resource in auth URL and token exchange
if neither is set:
    derive resource from gateway URL (existing behavior)
```

---

### Phase 1: Core Implementation

- [ ] **Add `audience` parameter to authorization URL** (`oauth_manager.py:~1143-1158`)
  - In `_create_authorization_url_with_pkce()`, check for `credentials.get("audience")`
  - Add `audience` to params dict when present
  - Support both `audience` and `resource` simultaneously

- [ ] **Add `audience` parameter to token exchange** (`oauth_manager.py:~1180-1210`)
  - In `_exchange_code_for_tokens()`, check for `credentials.get("audience")`
  - Add `audience` to token_data dict when present

- [ ] **Add `audience` to client_credentials flow** (`oauth_manager.py`)
  - Find `_get_client_credentials_token()` and add `audience` support

- [ ] **Add `audience` to token refresh flow** (`oauth_manager.py`)
  - Ensure `audience` is included when refreshing tokens

- [ ] **Update `_should_include_resource_parameter()`** (`oauth_manager.py:~1116-1119`)
  - If `audience` is set and `resource` is not, skip `resource`
  - If both are set, include both

### Phase 2: Schema & UI

- [ ] **Document `audience` in schema descriptions** (`schemas.py`)
  - Update `oauth_config` field descriptions at lines ~2525, ~2869, ~3199
  - Add `audience` to the documented keys

- [ ] **Add `audience` field to Admin UI OAuth form** (`admin.py`)
  - Add input field for `audience` in the gateway OAuth configuration section
  - Pre-populate from existing oauth_config

### Phase 3: Testing

- [ ] **Unit tests**: Test authorization URL includes `audience` when configured
- [ ] **Unit tests**: Test token exchange includes `audience`
- [ ] **Unit tests**: Test `audience`-only config omits `resource`
- [ ] **Unit tests**: Test both `audience` and `resource` together
- [ ] **Integration test**: End-to-end OAuth flow with `audience` parameter
- [ ] **Deny-path test**: Verify `audience` is not sent when not configured (backward compat)

### Phase 4: Documentation

- [ ] Update OAuth configuration docs with `audience` parameter
- [ ] Add Atlassian integration example to docs
- [ ] Add Auth0 integration example to docs

---

## ✅ Success Criteria

- [ ] Atlassian MCP server can be authorized through the gateway
- [ ] Auth0 audience-scoped tokens work correctly
- [ ] Existing OAuth flows (Entra, Keycloak, Okta) are unaffected
- [ ] Admin UI shows `audience` field in OAuth configuration
- [ ] Unit and integration tests pass

---

## 🔗 Related

- Discussion #3014 — original user report (Atlassian MCP)
- Discussion #3048 — external JWKS_URI validation (related OAuth topic)
- `mcpgateway/services/oauth_manager.py` — primary implementation file
- `docs/docs/architecture/oauth-design.md` — OAuth architecture documentation

### PR DESCRIPTION
**Pull request by bogdanmariusc10**
## 🔗 Related Issue

Closes #3499

## 📝 Summary

This PR adds support for the OAuth `audience` parameter to enable integration with Atlassian, Auth0, and other OAuth providers that use `audience` instead of (or in addition to) the RFC 8707 `resource` parameter.

**Problem**: Atlassian's OAuth endpoint requires `audience=api.atlassian.com`, but ContextForge only sent the `resource` parameter, blocking Atlassian MCP server integration.

**Solution**: Added optional `audience` parameter to OAuth configuration that:

- Works alongside existing `resource` parameter (both can be used simultaneously)
- Applies to all OAuth grant types (authorization_code, client_credentials, refresh_token)
- Maintains full backward compatibility with existing OAuth flows

### Core OAuth Manager
- **Added audience to authorization URL** (line 1491-1492): Includes `audience` parameter in OAuth authorization requests
- **Added audience to token exchange** (line 1568-1569): Includes `audience` in authorization code token exchange
- **Added audience to client credentials flow** (line 542-543): Includes `audience` in client credentials grant
- **Added audience to token refresh** (line 1679-1680): Includes `audience` in refresh token requests
- **Added validation helper** (line 174-203): `_validate_and_extract_audience()` validates format, strips whitespace, enforces length limits, and uses `sanitize_for_log()` for security
- **Added debug logging**: All 4 OAuth flow callsites log audience parameter with `sanitize_for_log(audience)` for troubleshooting

### Admin UI Integration
- **Added oauth_audience field extraction** in 4 form handlers:
  - Gateway create/edit (lines 12265, 12549)
  - A2A agent create/edit (lines 15739, 16009)
- **All handlers apply `.strip()`** to prevent whitespace-padded values from bypassing validation
- Fields are optional and only included when provided

### Schema Documentation
- **Updated GatewayCreate/Update/Read** oauth_config descriptions to document `audience` parameter
- **Updated A2AAgentCreate/Update/Read** oauth_config descriptions to document `audience` parameter
- All schemas now clearly indicate `audience` is supported for Atlassian/Auth0 integration

### User Documentation
- Added Atlassian and Auth0 integration examples
- Documented audience parameter usage and validation rules

## 🏷️ Type of Change

- ✅ Feature / Enhancement
- ✅ Documentation

## 🧪 Verification

| Check | Command | Status |
|-------|---------|--------|
| Lint suite | `make lint` | ✅ Pass |
| Unit tests | `make test` | ✅ Pass (17 new tests) |
| Coverage ≥ 80% | `make coverage` | ✅ Pass |

### Test Coverage

**11 tests in test_oauth_audience.py** covering:
- Authorization URL generation with/without audience
- Token exchange with audience parameter
- Client credentials flow with audience
- Token refresh with audience
- Audience + resource coexistence
- Validation edge cases (whitespace, empty, invalid chars, length limits)

**6 tests in test_admin.py** covering:
- Gateway create/edit forms with audience
- Server create/edit forms with audience  
- A2A agent create/edit forms with audience

All tests verify `audience` parameter is correctly included in OAuth requests and properly validated.

## ✅ Checklist

- ✅ Code formatted (`make black isort pre-commit`)
- ✅ Tests added/updated for changes
- ✅ Documentation updated
- ✅ No secrets or credentials committed

### OAuth Manager Changes (oauth_manager.py)

**Validation Pattern** (lines 42-44, 174-203):
- `_AUDIENCE_PATTERN`: Regex for URI/hostname validation (alphanumeric, dots, hyphens, underscores, colons, slashes)
- `_AUDIENCE_MAX_LENGTH`: 512 character limit
- `_validate_and_extract_audience()`: Centralized validation with `.strip()`, format check, length enforcement, and `sanitize_for_log()` error messages

**OAuth Flow Integration**:
- Line 1491-1492: Authorization URL includes audience
- Line 1568-1569: Token exchange includes audience
- Line 542-543: Client credentials includes audience
- Line 1679-1680: Token refresh includes audience
- All callsites include debug logging with `sanitize_for_log(audience)`

### Admin UI Integration (admin.py)

**Form Handler Pattern** (4 locations):
```python
oauth_audience = str(form.get("oauth_audience", "")).strip()
```

Applied consistently in:
- `admin_add_gateway` (line 12265)
- `admin_edit_gateway` (line 12549)
- `admin_add_a2a_agent` (line 15739)
- `admin_edit_a2a_agent` (line 16009)

### Backward Compatibility

- Existing OAuth flows without `audience` continue to work unchanged
- `resource` parameter behavior is unaffected
- Both `audience` and `resource` can be used together
- Validation only applies when `audience` is provided

### Example Configuration

```yaml
oauth_config:
  grant_type: authorization_code
  client_id: "my-atlassian-app"
  authorization_url: "https://auth.atlassian.com/authorize"
  token_url: "https://auth.atlassian.com/oauth/token"
  audience: "api.atlassian.com"
  scopes: ["read:jira-work", "write:jira-work"]
```

### Round 1 Fixes (Original Review)
- ✅ **B1**: Added `.strip()` to `admin_add_gateway` (line 12265)
- ✅ **B2**: Added input validation via `_validate_and_extract_audience()` with regex pattern and length limit
- ✅ **I4**: Added debug logging at all 4 audience callsites with `sanitize_for_log(audience)`
- ✅ **Mn1**: Eliminated 4-instance code duplication with single validation helper
- ✅ **Mn2**: Fixed HTML label associations for WCAG 2.1 AA compliance
- ✅ **T7**: Added test for audience + resource coexistence
- ✅ **T8**: Added 9 edge case validation tests in 340-line `test_oauth_audience.py`
- ✅ **I5**: Updated Gateway schema descriptions to document audience

### Round 2 Fixes (Follow-up Review)
- ✅ **New inconsistency**: Added `.strip()` to three remaining form handlers:
  - `admin_edit_gateway` (line 12549)
  - `admin_add_a2a_agent` (line 15739)
  - `admin_edit_a2a_agent` (line 16009)
- ✅ **I5 (A2A schemas)**: Verified A2AAgentCreate/Update/Read already document audience (lines 4649, 4978, 5316)

### Notes on Review Findings

**B3 - PR description claim**: The original PR description mentioned "Removed unreachable code in `_should_include_resource_parameter()`" but the diff shows +20 additions, 0 deletions. This was a documentation error - no code was removed from that function, only audience support was added.

**I6 - Resource suppression phrasing**: The original PR description implied audience suppresses resource ("instead of"). The actual behavior is that both parameters work simultaneously. The implementation correctly sends both when configured.

## 🎯 Testing Strategy

- **Unit tests** verify audience is included in all OAuth request types
- **Tests cover** both audience-only and audience + resource scenarios
- **Deny-path tests** ensure backward compatibility (no audience when not configured)
- **Provider-specific tests** for Atlassian and Auth0 use cases
- **Validation tests** cover edge cases: whitespace, empty strings, invalid characters, length limits

### CODE REVIEW

**Review by msureshkumar88 — CHANGES_REQUESTED** [review]
Thank you for this clear, well-scoped PR — the feature concept is solid and the coverage across all four grant types is appreciated. A few issues need addressing before this is ready to merge.

---

## Blocking

**1. Missing `.strip()` in gateway form handler — `admin.py:12244`**

Server and A2A handlers correctly call `.strip()`:
```python
oauth_audience = str(form.get("oauth_audience", "")).strip()  # ✅ server handlers (lines 2892, 3054)
```

But the gateway handler does not:
```python
oauth_audience = str(form.get("oauth_audience", ""))  # ❌ line 12244 — missing .strip()
```

A whitespace-padded audience value will propagate to the OAuth request and cause silent auth failures against Atlassian and Auth0. Worth fixing so all handlers behave consistently.

---

**2. No input validation on `audience` value — `oauth_manager.py:503, 1451, 1527, 1637`**

`audience` flows from form input through `oauth_config` directly into outbound OAuth request parameters with no format validation. An admin-level misconfiguration (or a compromised admin session) can pass values containing embedded control characters or whitespace. The codebase already has `sanitize_for_log()` for log injection prevention — the same discipline should apply here.

Per OAuth spec, `audience` must be a URI or hostname. Consider at minimum:
- Non-empty after strip
- URI/hostname format check (e.g. `re.match(r'^[a-zA-Z0-9][a-zA-Z0-9\.\-_/:]*$', value)`)
- Reasonable max length bound

Reject on validation failure rather than silently forwarding the value to the provider.

---

**3. Commit message and PR notes misrepresent the diff**

PR notes state: *"Removed unreachable code in `_should_include_resource_parameter()`"*. However, `oauth_manager.py` shows **+20 additions, 0 deletions** — no code was removed from that function. Either the removal was accidentally dropped, or the description needs correcting. Creates a misleading audit trail either way.

---

## Important

**4. No debug logging when `audience` is applied — all 4 callsites**

Every other credential decision in the token request path has an accompanying `logger.debug()` call. The four `audience` insertion blocks have none. When Atlassian or Auth0 integration fails, there's no log evidence confirming whether `audience` was sent. Suggest adding:
```python
logger.debug("Including audience parameter in OAuth request: %s", sanitize_for_log(audience))
```

---

**5. `A2AAgentUpdate` and `A2AAgentRead` schema descriptions not updated — `schemas.py`**

`oauth_config` descriptions were updated in `GatewayCreate`, `GatewayUpdate`, `GatewayRead`, and `A2AAgentCreate` to document the `audience` key. `A2AAgentUpdate` and `A2AAgentRead` were missed. API consumers reading those schemas will not know `audience` is a valid key for A2A agent updates.

---

**6. PR description claim about `resource` suppression is misleading**

*"When set without `resource`, the RFC 8707 `resource` parameter is automatically omitted"* — this is not a new behavior introduced by this PR. `resource` is omitted whenever `_should_include_resource_parameter()` returns `False`, independent of whether `audience` is set. The claim should be removed or corrected to avoid confusion for future maintainers.

---

## Testing Gaps

**7. No test for `audience` + `resource` coexistence in the same outbound request**

The PR claims both can coexist simultaneously, but `test_oauth_audience.py` only covers `audience`-only and `audience`-absent cases. A test verifying that when both `audience` and `resource` are configured, both appear in the serialized form body would close this gap (and is the scenario Auth0 + RFC 8707-compliant servers would use).

---

**8. No negative / validation tests**

No tests for: whitespace-only `audience`, empty string after strip, overlong value, or control-character injection. Given that validation is currently absent (finding #2), these tests would fail today — which is exactly why they're worth adding alongside the fix.

---

**9. Gateway form handler coverage unclear — `test_admin.py`**

PR description states "4 form handlers" but separately lists "gateway create/edit, server create/edit, A2A agent create/edit" (6 operations). Grep across `admin.py` found 3 distinct `oauth_audience` extraction points. Worth confirming which specific handlers have test coverage, since the gateway handler also has the missing-strip bug (finding #1).

---

**10. No integration-level test verifying actual HTTP request body**

All tests mock at the service layer. No test fires a request to a mock OAuth endpoint and asserts the `application/x-www-form-urlencoded` body actually contains `audience=api.atlassian.com`. This matters because a type mismatch in the payload dict could silently drop the param through `urlencode`.

---

## Minor

**11. Duplicate `audience` extraction pattern across 4–5 sites in `oauth_manager.py`**

```python
audience = credentials.get("audience")
if audience:
    token_data["audience"] = audience
```

This block is repeated 4–5 times with only the source dict name changing. A small helper `_apply_audience(source, payload)` would centralize any future fix (e.g., adding validation from finding #2).

---

**12. Admin HTML inputs missing `<label>` association — `admin.html:5985, 7440`**

Both `oauth_audience` fields use a `<p>` tag as description rather than a properly associated `<label for="...">`. Screen readers won't announce the field purpose (WCAG 2.1 AA, 1.3.1).

---

## Summary

| Category | Count |
|---|---|
| Blocking | 3 |
| Important | 3 |
| Testing gaps | 4 |
| Minor | 2 |

The core implementation is clean and the grant-type coverage is complete. Resolving the three blocking items (missing strip, absent input validation, inaccurate commit description) and the testing gaps will get this over the line.

**Comment by bogdanmariusc10** [comment]
Hello, @msureshkumar88 ! I have addressed all of your review findings in commit [3aa46d4](https://github.com/IBM/mcp-context-forge/pull/4795/commits/3aa46d42ec0ea6c9a6c0db04c1cc4bfa87bb59a7).

**Review by msureshkumar88 — CHANGES_REQUESTED** [review]
The fixes address the majority of the original findings — the validation helper is well-constructed, debug logging is consistent, and the test coverage is now solid. One new issue was introduced during the fix that needs addressing before merge.

---

## ✅ What's Fixed

| Finding | Resolution |
|---|---|
| B1: Missing `.strip()` in `admin_add_gateway` | Fixed — `.strip()` added at line ~12265 |
| B2: No input validation on `audience` | Fixed — `_validate_and_extract_audience()` with `_AUDIENCE_PATTERN` regex + `_AUDIENCE_MAX_LENGTH=512` |
| I4: No debug logging at 4 audience callsites | Fixed — all 4 callsites log with `sanitize_for_log(audience)` |
| Mn1: 4-instance code duplication | Fixed — single helper eliminates repetition |
| Mn2: Missing HTML label associations | Fixed — WCAG 2.1 AA |
| T7: No audience+resource coexistence test | Fixed |
| T8: No negative/validation tests | Fixed — 9 edge case tests + 340-line `test_oauth_audience.py` |
| I5 (partial): Gateway schema descriptions | Fixed — `GatewayCreate/Update/Read` now document `audience` |

The `_validate_and_extract_audience()` helper is solid: strips whitespace, applies the URI/hostname regex, enforces a length bound, and uses `sanitize_for_log()` in error messages. Good pattern.

---

## 🔴 Blocking

**New inconsistency: `.strip()` applied to `admin_add_gateway` but not to the other three handlers**

The fix correctly added `.strip()` to `admin_add_gateway`, but the same fix commit also introduced audience support to `admin_edit_gateway`, `admin_add_a2a_agent`, and `admin_edit_a2a_agent` — without the same strip:

```python
# admin_add_gateway (~line 12265) — fixed ✅
oauth_audience = str(form.get("oauth_audience", "")).strip()

# admin_edit_gateway (~line 12546) — missing ❌
oauth_audience = str(form.get("oauth_audience", ""))

# admin_add_a2a_agent (~line 15736) — missing ❌
oauth_audience = str(form.get("oauth_audience", ""))

# admin_edit_a2a_agent (~line 16006) — missing ❌ (same pattern)
```

An admin editing an existing gateway or managing A2A agents can submit a whitespace-padded audience value that bypasses the strip and flows into `oauth_config` unstripped. The validation happens later in `_validate_and_extract_audience()` at token-request time, not at form-submission time — so the inconsistency creates a silent window where bad data gets stored. Worth making all four handlers consistent.

---

## Notes (non-blocking)

**B3 — PR description dead-code claim still present**
The commit message says "Removed unreachable code in `_should_include_resource_parameter()`" but the diff shows +20 additions, 0 deletions in that function. No code was removed. Worth correcting the PR description to avoid a misleading audit trail.

**I5 (partial) — A2A agent schemas not updated**
`GatewayCreate/Update/Read` oauth_config descriptions now mention `audience` — good. `A2AAgentCreate`, `A2AAgentUpdate`, and `A2AAgentRead` were not updated. API consumers querying A2A agent schemas get no hint that `audience` is a supported field. Worth a one-liner addition to each.

**I6 — Resource suppression phrasing**
PR description still implies `audience` suppresses `resource`. The code correctly sends both simultaneously. A small wording change — "works alongside `resource`" rather than "instead of" — would match the actual behavior.

**Comment by bogdanmariusc10** [comment]
@msureshkumar88 Added `.strip()` to `admin_edit_gateway`, `admin_add_a2a_agent`, and `admin_edit_a2a_agent` and updated the PR description.

**Review by msureshkumar88 — DISMISSED** [review]
All blocking and important findings from both review rounds are resolved in the current HEAD.

**Round 1 fixes confirmed:**
- B1: `.strip()` added to `admin_add_gateway`
- B2: `_validate_and_extract_audience()` with URI/hostname regex and 512-char limit
- I4: Debug logging at all 4 callsites via `sanitize_for_log()`
- Mn1: Code duplication eliminated via single helper
- Mn2: HTML `<label>` associations added (WCAG 2.1 AA)
- T7/T8: Audience+resource coexistence test and 9 negative/validation tests added

**Round 2 blocking fix confirmed (295e8d6d):**
- `.strip()` now applied consistently across all four handlers: `admin_add_gateway`, `admin_edit_gateway`, `admin_add_a2a_agent`, `admin_edit_a2a_agent`. Surgical 3-line change, no side effects.

**Schemas complete:**
- `A2AAgentCreate`, `A2AAgentUpdate`, and `A2AAgentRead` all updated with `audience` in their `oauth_config` descriptions.

**PR description corrected:**
- B3 acknowledged (no code removed from `_should_include_resource_parameter()`, was a documentation error)
- I6 fixed — description now accurately states both parameters work simultaneously

One non-blocking gap remains from the original review (T10: no integration-level test asserting the serialized HTTP body contains `audience`). This was not re-raised as blocking in Round 2 and the `_validate_and_extract_audience()` helper mitigates the risk. Suggest tracking as a follow-up rather than blocking merge.

**Comment by msureshkumar88** [comment]
Hi @bogdanmariusc10 — there's a merge conflict in `mcpgateway/admin.py`. Could you rebase from `main` to resolve it? Once that's done I can re-approve.

```bash
git fetch upstream
git rebase upstream/main
# resolve conflicts in mcpgateway/admin.py, then:
git add mcpgateway/admin.py
git rebase --continue
git push --force-with-lease
```

Thanks!

**Comment by ja8zyjits** [comment]
Please rebase.

**Comment by bogdanmariusc10** [comment]
Hello, @ja8zyjits ! Successfully rebased this branch.

File: tests/unit/mcpgateway/services/test_oauth_audience.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+                "audience": "https://my-api.example.com",
+                "resource": "https://mcp-server.example.com",
+                "scopes": ["read"],
+            }
+
+            result = await oauth_manager._exchange_code_for_tokens(credentials=credentials, code="auth-code", code_verifier="verifier")
+
+            # Verify both parameters are in the request
+            oauth_manager._post_token_request.assert_called_once()
+            call_args = oauth_manager._post_token_request.call_args
```
Comment (ja8zyjits) [inline tests/unit/mcpgateway/services/test_oauth_audience.py:330]: These calls are duplicated below?
```python
call_args = oauth_manager._post_token_request.call_args
token_data = call_args[0][1]
assert token_data["audience"] == "https://my-api.example.com"
```

**Review by ja8zyjits — CHANGES_REQUESTED** [review]

**Comment by ja8zyjits** [comment]
HI @bogdanmariusc10 
Please remove that duplicate and fix the secrets, I will quque it right away, thanks.

**Comment by bogdanmariusc10** [comment]
Hi, @ja8zyjits ! Removed the duplicate code and fixed .secrets.baseline.

**Review by ja8zyjits — APPROVED** [review]
LGTM
